cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java

```diff
@@ -139,180 +162,219 @@ private boolean throttlingEnabled() {
+        continue;
+      }
 
-    // Merge new throttled replicas with existing configuration values.
-    Set<String> newThrottledReplicas = new TreeSet<>(replicas);
-    if (currThrottledReplicas != null) {
-      newThrottledReplicas.addAll(Arrays.asList(currThrottledReplicas.split(",")));
+      // Merge new throttled replicas with existing configuration values.
+      Set<String> newThrottledReplicas = new TreeSet<>(replicas);
+      if (currThrottledReplicas != null && !currThrottledReplicas.value().equals("")) {
+        newThrottledReplicas.addAll(Arrays.asList(currThrottledReplicas.value().split(",")));
+      }
+      ops.add(new AlterConfigOp(new ConfigEntry(replicaThrottleConfigKey, String.join(",", newThrottledReplicas)), AlterConfigOp.OpType.SET));
+    }
+    if (!ops.isEmpty()) {
+      changeTopicConfigs(topic, ops);
     }
-    topicConfigs.setProperty(replicaThrottleConfigKey, String.join(",", newThrottledReplicas));
-    changeTopicConfigs(topic, topicConfigs);
   }
 
-  /**
-   * Update topic configuration properties if a topic exist. No-op for non-existent topic
-   *
-   * @param topicName topic name
-   * @param topicConfigs configuration properties to update
-   */
-  private void changeTopicConfigs(String topicName, Properties topicConfigs) {
+  void changeTopicConfigs(String topic, Collection<AlterConfigOp> ops)
+  throws ExecutionException, InterruptedException, TimeoutException {
+    ConfigResource cf = new ConfigResource(ConfigResource.Type.TOPIC, topic);
+    Map<ConfigResource, Collection<AlterConfigOp>> configs = Collections.singletonMap(cf, ops);
     try {
-      ExecutorUtils.changeTopicConfig(_adminZkClient, topicName, topicConfigs);
+      _adminClient.incrementalAlterConfigs(configs).all()
+          .get(CLIENT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      waitForConfigs(cf, ops);
     } catch (Exception e) {
-      if (!_kafkaZkClient.topicExists(topicName)) {
-        LOG.debug("Failed to change configs for topic {} since it does not exist", topicName);
+      if (!topicExists(topic)) {
+        LOG.debug("Failed to change configs for topic {} since it does not exist", topic);
         return;
       }
       throw e;
     }
   }
 
+  void changeBrokerConfigs(int brokerId, Collection<AlterConfigOp> ops)
+  throws ExecutionException, InterruptedException, TimeoutException {
+    ConfigResource cf = new ConfigResource(ConfigResource.Type.BROKER, String.valueOf(brokerId));
+    Map<ConfigResource, Collection<AlterConfigOp>> configs = Collections.singletonMap(cf, ops);
+    _adminClient.incrementalAlterConfigs(configs).all()
+        .get(CLIENT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    waitForConfigs(cf, ops);
+  }
+
+  boolean topicExists(String topic) throws InterruptedException, TimeoutException {
+    try {
+      return _adminClient.listTopics().names().get(CLIENT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS).contains(topic);
+    } catch (ExecutionException ee) {
+      return false;
```

Comment:
I am thinking about what the right response should be in case there is an `ExecutionException` while listing topics.
I wonder if it would be safer to log the error and throw the exception as we did for `InterruptedException` and `TimeoutException`?

Reply:
That's a good point, I've pushed an update